File: src/main/storage/service-file.ts
// service.toml — Periyodik Bakım, the PERİYODİK BAKIM sheet's shape
// (XTRITIUM §4.4, §6.2).
//
// `vendor` holds a pasted address. It is selectable text only and never a link
// (§3.5): nothing in TRITIUM opens a browser, so the storage layer treats it as
// the plain string it is and asks no questions about its contents.

import type { ServiceEntry } from '../../shared/records.js'
import { formatMoney, toMoney } from '../../shared/scaled.js'
import {
  emptyDocument,
  readEntryFile,
  writeEntryFile,
  type EntryDocument,
  type EntrySpec
} from './entry-file.js'
import { basicString, dateLines, line, readDate, readInteger, readNumber, readString } from './toml.js'

const KNOWN_KEYS = ['id', 'date', 'part', 'odometer_km', 'amount', 'vendor'] as const

export const SERVICE_SPEC: EntrySpec<ServiceEntry> = {
  kind: 'service',
  knownKeys: KNOWN_KEYS,
  readEntry: (raw, id) => ({
    id,
    date: readDate(raw['date']),
    part: readString(raw['part']),
    odometer_km: readInteger(raw['odometer_km']),
    amount: toMoney(readNumber(raw['amount'])),
    vendor: readString(raw['vendor'])
  }),
  emitEntry: (entry) => [
    line('id', basicString(entry.id)),
    ...dateLines('date', entry.date),
    line('part', basicString(entry.part)),
    line('odometer_km', entry.odometer_km.toString()),
    line('amount', formatMoney(entry.amount)),
    line('vendor', basicString(entry.vendor))
  ]
}

export type ServiceDocument = EntryDocument<ServiceEntry>

export function emptyService(): ServiceDocument {
  return emptyDocument<ServiceEntry>()
}

export function readService(file: string): ServiceDocument {
  return readEntryFile(file, SERVICE_SPEC)
}

export function writeService(file: string, document: ServiceDocument): void {
  writeEntryFile(file, document, SERVICE_SPEC)
}

File: android/app/src/main/kotlin/io/github/sudomegas/tritium/storage/ServiceFile.kt
package io.github.sudomegas.tritium.storage

import dev.eav.tomlkt.TomlTable
import java.io.File

/** `service.toml` — the Periyodik Bakım sheet's shape (XTRITIUM §4.4), ported from `service-file.ts`. */

private val KNOWN_KEYS = setOf("id", "date", "part", "odometer_km", "amount", "vendor")

object ServiceSpec : EntrySpec<ServiceEntry> {
    override val kind = RecordKind.SERVICE
    override val knownKeys = KNOWN_KEYS

    override fun readEntry(table: TomlTable, id: String): ServiceEntry = ServiceEntry(
        id = id,
        date = readDate(table, "date"),
        part = readString(table, "part"),
        odometerKm = readInteger(table, "odometer_km"),
        amount = Scaled.toMoney(readNumber(table, "amount")),
        // A pasted address: selectable text ONLY, never a link (XTRITIUM §5)
        // — enforced by the UI (AF6), not by the storage layer, which just
        // carries the string.
        vendor = readString(table, "vendor"),
    )

    override fun emitEntryFields(entry: ServiceEntry): List<String> = buildList {
        addAll(dateLines("date", entry.date))
        add(line("part", basicString(entry.part)))
        add(line("odometer_km", entry.odometerKm.toString()))
        add(line("amount", Scaled.formatMoney(entry.amount)))
        add(line("vendor", basicString(entry.vendor)))
    }
}

typealias ServiceDocument = EntryDocument<ServiceEntry>

fun emptyService(): ServiceDocument = emptyDocument()
fun readService(file: File): ServiceDocument = readEntryFile(file, ServiceSpec)
fun writeService(file: File, document: ServiceDocument) = writeEntryFile(file, document, ServiceSpec)
